### DISCUSSION TITLE
Obtain previous visited URLs

### DISCUSSION THREAD

**Discussion by captain-yossarian** [vercel/next.js#39502] [Help] [answered]
Hi, I need to get previous visited URL in nextjs/router. Please see my [question on stackoverflow](https://stackoverflow.com/questions/73321217/get-previous-two-urls-from-next-router-object).

So, `useRouter` returns `NextRouter`. 
`NextRouter` type does not contain `components` property whereas runtime value contains it. This property (`components`) contains previously visited URLS as an object keys, however, they are unordered. Is it possible to get exact URL which was visited previously?
I don't need to call `router.back()`

Is it possible to emit custom router events ?

**Comment by gaston-flores** [comment, marked as answer]
I don't think Next has anything like this built in. The `components` property you found seems to be something for internal use, and is not documented anywhere.

You could use [`events`](https://nextjs.org/docs/api-reference/next/router#usage-6) to listen for route changes and build a history that way. 

```js
  const router = useRouter();
  const [history, setHistory] = useState([router.asPath]);

  useEffect(() => {
    const handleRouteChange = (url) => {
      setHistory((history) => [...history, url]);
    };

    router.events.on("routeChangeComplete", handleRouteChange);
    return () => {
      router.events.off("routeChangeComplete", handleRouteChange);
    };
  }, [router.events]);

  console.log(history);
```

Here's [an example](https://codesandbox.io/s/kind-fog-yxxewd?file=/pages/_app.js). There's a few caveats:

- You need to put this in `_app` to avoid losing the history when the route changes, use Context if you need to pass it to your pages.
- Your history will be lost if the user reloads the page.

And probably others I haven't noticed. Are you sure you need to do this?
